src/ansi.rs

```diff
@@ -0,0 +1,1849 @@
+        b: color_from_slice(&color[item_len * 2..])?,
+    })
+}
+
+fn parse_number(input: &[u8]) -> Option<u8> {
+    if input.is_empty() {
+        return None;
+    }
+    let mut num: u8 = 0;
+    for c in input {
+        let c = *c as char;
+        if let Some(digit) = c.to_digit(10) {
+            num = match num.checked_mul(10).and_then(|v| v.checked_add(digit as u8)) {
+                Some(v) => v,
+                None => return None,
+            }
+        } else {
+            return None;
+        }
+    }
+    Some(num)
+}
+
+/// Internal state for VTE processor.
+struct ProcessorState<T: TimeProvider> {
+    /// Last processed character for repetition.
+    preceding_char: Option<char>,
+
+    /// DCS sequence waiting for termination.
+    #[cfg(feature = "alloc")]
+    dcs: Option<Dcs>,
+
+    /// State for synchronized terminal updates.
+    #[cfg(feature = "alloc")]
+    sync_state: SyncState<T>,
+
+    _time_provider: PhantomData<T>,
+}
+
+/// Current timestamp provider
+pub trait TimeProvider {
+    /// Get current timestamp from unix epoch in milliseconds
+    fn now(&self) -> u128;
+}
+
+/// Time provider which always returns 0
+#[derive(Default)]
+pub struct NullTimeProvider {}
+
+impl TimeProvider for NullTimeProvider {
+    fn now(&self) -> u128 {
+        0
+    }
+}
+
+#[derive(Debug)]
+#[cfg(feature = "alloc")]
+struct SyncState<T: TimeProvider> {
+    /// Expiration time (in milliseconds since unix epoch) of the synchronized update.
+    timeout: Option<u128>,
+
+    in_sync: bool,
+
+    /// Sync DCS waiting for termination sequence.
+    pending_dcs: Option<Dcs>,
+
+    /// Bytes read during the synchronized update.
+    buffer: Vec<u8>,
+
+    /// Time provider
+    time_provider: T,
+}
+
+#[cfg(feature = "alloc")]
+impl<T: TimeProvider> SyncState<T> {
+    fn new(time_provider: T) -> SyncState<T> {
+        Self {
+            buffer: Vec::with_capacity(SYNC_BUFFER_SIZE),
+            pending_dcs: None,
+            timeout: None,
+            in_sync: false,
+            time_provider,
+        }
+    }
+}
+
+/// Pending DCS sequence.
+#[derive(Debug)]
+#[cfg(feature = "alloc")]
+enum Dcs {
+    /// Begin of the synchronized update.
+    SyncStart,
+
+    /// End of the synchronized update.
+    SyncEnd,
+}
+
+/// The processor wraps a `Parser` to ultimately call methods on a Handler.
+pub struct Processor<T: TimeProvider> {
+    state: ProcessorState<T>,
+    parser: Parser,
+}
+
+impl Processor<NullTimeProvider> {
+    #[cfg(feature = "alloc")]
+    pub fn new() -> Processor<NullTimeProvider> {
+        Processor::new_with_time_provider(NullTimeProvider::default())
+    }
+
+    #[cfg(not(feature = "alloc"))]
+    pub fn new() -> Processor<NullTimeProvider> {
+        Processor {
+            parser: Parser::new(),
+            state: ProcessorState {
+                preceding_char: None,
+                _time_provider: PhantomData::<NullTimeProvider>::default(),
+            },
+        }
+    }
+}
```

Comment:
I don't think providing these is particularly beneficial. Without alloc a time provider shouldn't even be necessary in the first place.

Reply:
I agree with you, but how do I remove `TimeProvider` generic parameter from `impl Perform for Performer` in no-alloc environment?